### ISSUE TITLE
Do you need any help with maintenance?

### ISSUE THREAD

**Issue by matthewdeanmartin** [njvack/bids-json-writer#14]
I accidentally wrote a python-to-markdown serializer/deserializer [markpickle](https://github.com/matthewdeanmartin/markpickle) before realizing that your code does the deserializing part much the same way or better & my ATX-headers-as-dict code now follows your pattern.

I sort of would like to fix the 10 issues and 3 pull requests so I can validate that my serializer/deserializer can handle the same cases that markdown-to-json can handle & make that part of the build process.

I could
- fork & publish to a new name 
- you could grant me collaborator or owner rights on pypi, with owner rights I can do a "trusted publisher" release, with collaborator I can only do username/password publication
- you could grant me rights on this github repo & pypi

Having done this a few time before...
I asked another dude if they needed help & they gave me github only rights & disappeared for a few years. Not useful! I'd have to still publish to pypi under a different name.
I asked yet another dude if they needed help & they gave me just pypi collaboration rights- good enough, I forked & could publish, but can't do the new 'trusted publisher thing'
Another dude just didn't reply for years, so I forked and published under a similar name.

Anyhow, before I take that last option, I wanted to get your feedback.

**Comment by njvack** [comment]
Thank you! Clearly I do need help with maintaining this, because boy howdy have I not done any of that myself for a long time.

I think giving you commit rights here and owner rights on pypi makes sense. My only worry is that, like, you're secretly going to change this to be a malware package. One that would affect like eighteen people, heh. Probably not a serious risk.

Anyhow! I think I can grant you commit rights here now, messing around in pypi is… something I do less often but it's probably straightforward enough…

**Comment by njvack** [comment]
Okay you should have collab rights here now.

Also! Maybe you have insight into this, but… why is this a useful project? I wrote it because I needed to write dataset descriptions in .json, and doing it in Markdown seemed easier. But I imagined this was such a niche use that no one else would ever want something like it.

**Comment by matthewdeanmartin** [comment]
Thanks!

>My only worry is that, like, you're secretly going to change this to be a malware package.

It is a huge problem. Maybe I should suggest to the pypi folk provided a better way to let people see that person A on pypi is the same person on github and that they have a reputation.

>Maybe you have insight into this, but… why is this a useful project?

For deserialization, real Document Object Models (DOMs) are pedantic and too detailed because they have to model everything you need to turn Markdown into HTML. Markdown-to-json turns most simple markdown into something that is as easy as json to work with, provided the user doesn't care about arbitrarily complex documents.

For the serialization side, API developers want to provide as many formats as possible to users of an API, so if people want XML, JSON, Bibtex we give it to them & don't worry about why they chose that format (XML was because someone made them). But if we give them HTML or microformats, now I have to worry about XSS attacks. But if I give people Markdown, it is much safer because it was designed with untrusted text (blog comments) in mind.

My markpickle library was original more to the serialization side, but what gets serialized should be deserialized to prove that it succeeded, so serialization, you also need to solve deserialization.

**Comment by matthewdeanmartin** [comment]
Okay, merged in
- github action
- devcontainer
- build script
- linted code, except for 
- added black
- started switch to pyproject.toml
- added readme.md to setup.py in the meanwhile
- more unit tests to process example files

I'll start working the 10 bugs next.

**Comment by njvack** [comment]
Thank you!

Also, who are you on pypi?

**Comment by matthewdeanmartin** [comment]
This is me: https://pypi.org/user/matthewdeanmartin/

If you could make me (co) owner or set up the trusted publisher:
https://docs.pypi.org/trusted-publishers/adding-a-publisher/ that would be
appreciated.

(I'm trying to get out of the business of using passwords for package
upload)

Mathew Martin

**Comment by matthewdeanmartin** [comment]
Hey, we caught up on all the issues & MRs. Save for publishing a 2.0.0 to pypi were good. I added the min python version and bumped the first digit of the version number, not sure what else we can do for people on 2.x that didn't pin previously.

**Comment by njvack** [comment]
Awesome, thank you. I've added you as an owner to the pypi project and done… something… to set a trusted publisher, but I don't know that I did it right. I've never tried going the "publish to pypi with github actions" route.

**Comment by njvack** [comment]
Also! If you feel like it, please add yourself to the credits for this, and make note of markpickle in the readme.

I don't anticipate needing to make many further changes to this project, as long as python doesn't drastically change.

At one point, I started working on switching to the parser in [markdown-it-py](https://github.com/executablebooks/markdown-it-py) but then realized it would be a bunch of work for, probably, no benefit.

I guess the one thing I might do is add a single call that will dictify a markdown string… needing to explicitly build the AST is not something people really need to do…

**Comment by njvack** [comment]
Added `markdown_to_json.dictify()` and `markdown_to_json.jsonify()`

**Comment by matthewdeanmartin** [comment]
Looks like the project is now maintained & I can do MRs & publish to pypi, so I'll close this issue.

Woot!

**Comment by njvack** [comment]
Nice! How does the trusted publisher thing work, anyhow? How do you invoke the pypi-publish action?

**Comment by matthewdeanmartin** [comment]
Trusted publisher means the package owner said "expect a package to be published from gitlhub" and when a magic (cryptography stuff) runs, pypi says, "Oh that is github, so we know what hardware & what security controls Github has, so that package is probably controlled by the same entity as the package owner".

I'm still learning on how to do it, so I don't know how to get it to trigger only on a tag. On my other projects, I set the github environment (bad name), which in this case is pypi-publisher... I set it to "manual approve" and then tell the github action to always publish (both tag & non tag). If it runs an extra time, it would be rejected as already uploaded.

If it ran on a fork or a branch, pypi would see it and say, "that doesn't match the things that the package owner set" and it would reject.
